src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -589,13 +589,27 @@ private void InitializePackagesFolderInfo(string packagesFolderPath, bool exclud
             // We have been given the exact PackageIdentities (id and version) to update to e.g. from PMC update-package -Id <id> -Version <version>
             if (packageIdentities.Count > 0)
             {
-                primaryTargets = packageIdentities;
+                primaryTargets = new List<PackageIdentity>();
                 primaryTargetIds = packageIdentities.Select(p => p.Id);
 
                 // If we have been given explicit PackageIdentities to install then we will naturally prefer that
                 foreach (var packageIdentity in packageIdentities)
                 {
-                    preferredVersions[packageIdentity.Id] = packageIdentity;
+                    // Just a check to make sure the preferredVersions created from the existing package list actually contains the target
+                    if (preferredVersions.ContainsKey(packageIdentity.Id))
+                    {
+                        // If there was a version specified we will prefer that version
+                        if (packageIdentity.HasVersion)
+                        {
+                            preferredVersions[packageIdentity.Id] = packageIdentity;
+                            ((List<PackageIdentity>)primaryTargets).Add(packageIdentity);
```

Comment:
Could this just make a new list as the top instead of starting as IEnumerable? This just looks crazy.